phoenix-core/src/it/java/org/apache/phoenix/query/MaxConcurrentConnectionsIT.java

```diff
@@ -0,0 +1,152 @@
+import static org.apache.phoenix.util.PhoenixRuntime.JDBC_PROTOCOL_SEPARATOR;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class MaxConcurrentConnectionsIT extends BaseUniqueNamesOwnClusterIT {
+
+    private static HBaseTestingUtility hbaseTestUtil;
+
+    @BeforeClass
+    public static void setUp() throws Exception {
+        hbaseTestUtil = new HBaseTestingUtility();
+        hbaseTestUtil.startMiniCluster(1,1,null,null,DelayedRegionServer.class);
+        // establish url and quorum. Need to use PhoenixDriver and not PhoenixTestDriver
+        String zkQuorum = "localhost:" + hbaseTestUtil.getZkCluster().getClientPort();
+        url = PhoenixRuntime.JDBC_PROTOCOL + JDBC_PROTOCOL_SEPARATOR + zkQuorum +
+                JDBC_PROTOCOL_SEPARATOR + "uniqueConn=A";
+        DriverManager.registerDriver(PhoenixDriver.INSTANCE);
+    }
+
+    private String getUniqueUrl() {
+        return url + generateUniqueName();
+    }
+
+    //Have to shutdown our special delayed region server
+    @AfterClass
+    public static void tearDown() throws Exception {
+        hbaseTestUtil.shutdownMiniCluster();
+    }
+
+    @Test
+    public void testCannotCreateOverMaximumConnections() throws Exception {
+        int maxConnections = 3;
+        Properties props = new Properties();
+        props.setProperty(CLIENT_CONNECTION_MAX_ALLOWED_CONNECTIONS,String.valueOf(maxConnections));
+        props.setProperty(RENEW_LEASE_ENABLED,String.valueOf(false));
+        assertEquals(0,GLOBAL_OPEN_PHOENIX_CONNECTIONS.getMetric().getValue());
+
+        List<Connection> connections = new ArrayList<>();
+        String url = getUniqueUrl();
+
+        try{
+            for(int i = 0; i < maxConnections; i++) {
+                connections.add(DriverManager.getConnection(url, props));
+            }
+
+            try {
+                DriverManager.getConnection(url, props);
+                fail();
+            } catch(SQLException e) {
+                assertEquals(NEW_CONNECTION_THROTTLED.getErrorCode(),e.getErrorCode());
+                assertEquals(NEW_CONNECTION_THROTTLED.getSQLState(),e.getSQLState());
+            }
+
+        } finally {
+            for(Connection conn : connections) {
+                conn.close();
+            }
+        }
+    }
+
+    /**
+     * This tests the delete path which creates a internal phoenix connection per region
+     * @throws Exception
+     */
+    @Test
+    public void testDeleteRuntimeFailureClosesConnections() throws Exception {
+        String tableName = generateUniqueName();
+
+        //table with lots of regions
+        String ddl = "create table " + tableName +  "  (i integer not null primary key, j integer) SALT_BUCKETS=256 ";
+
+        Properties props = new Properties();
+        props.setProperty(QueryServices.ENABLE_SERVER_SIDE_DELETE_MUTATIONS,
```

Comment:
What happens if server-side mutations are enabled? If enabled, since we no longer would need to `rs.next` and read all the data from the server + generate delete mutations for them, is there any difference in behavior w.r.t. whether we create cloned connections?

Reply:
Tested locally and does not appear to be a behavior change in terms of leaked internal phoenix connections.

Reply:
In that case, is setting this property necessary?